[wasm][debugger] Reuse debugger-agent on wasm debugger

thaystg opened the pull request:
This will remove a lot of code that does the same thing on mini-wasm-debugger.

The code for debugger before this PR was shared between mini-wasm-debugger, library_mono.js and the BrowserDebugProxy. With this new approach the important code for the debugger are mainly on debugger-agent and BrowserDebugProxy.

This PR affects the size of dotnet.wasm and dotnet.js as is described below.

<google-sheets-html-origin><!--td {border: 1px solid #ccc;}br {mso-data-placement:same-cell;}-->
&nbsp; | without PR | with PR | bytes | % | &nbsp;
-- | -- | -- | -- | -- | --
dotnet.js | 249786 | 225242 | 24544 | 9,83% | REDUCED
dotnet.wasm | 2277406 | 2299707 | 22300 | 0,98% | INCREASED
brotli dotnet.js | 52291 | 47638 | 4653 | 8,90% | REDUCED
brotli dotnet.wasm | 780756 | 793778 | 13022 | 1,67% | INCREASED

</google-sheets-html-origin>

thaystg committed:
Trying to reuse debugger-agent on wasm debugger. This will remove a lot of code that does the same thing on mini-wasm-debugger.

thaystg committed:
Replace remove_breakpoint and clear_all_breakpoints with the ones on debugger-agent and remove unused code.

thaystg committed:
Stepping and callstack using debugger-agent.

thaystg committed:
Remove more code.

thaystg committed:
Get frame values using debugger-agent.

lewing reviewed (commented):
Sharing this code will be a great step forward and the prototype is coming along nicely.  Nice work.

thaystg committed:
Working valuetypes and call function on valuetypes.
make -C src/mono/wasm/ run-debugger-tests TEST_FILTER=DebuggerTests.SteppingTests.InspectValueTypeMethodArgsWhileStepping is working without use_cfo.

thaystg committed:
Failed:   316, Passed:   175

thaystg committed:
Failed:   307, Passed:   184, Skipped:     0, Total:   491

thaystg committed:
Failed:   280, Passed:   211

thaystg committed:
Failed:   277, Passed:   214

thaystg committed:
Implemented boxed value.
Failed:   271, Passed:   220

radical commented on src/mono/mono/mini/debugger-agent.c:
```diff
@@ -5222,9 +5252,8 @@ decode_vtype (MonoType *t, MonoDomain *domain, gpointer void_addr, gpointer void
 	ErrorCode err;
 
 	is_enum = decode_byte (buf, &buf, limit);
-	/* Enums are sent as a normal vtype */
-	if (is_enum)
-		return ERR_NOT_IMPLEMENTED;
+	if (CHECK_PROTOCOL_VERSION(2, 61))
+		decode_byte (buf, &buf, limit);
```
What happens in versions other than `2.61` now?

radical reviewed (commented):
Just posting some comments on the bits that I was able to look at (very little!).

thaystg committed:
Implementing get properties on objects.
Implementing handling error on debugger-agent.

thaystg committed:
Implementing callfunctionon object.
Failed:   248, Passed:   243

thaystg committed:
Implementing get pointer values.
Failed:   243, Passed:   248

thaystg committed:
Fixing pointer values and implement call on function with pointers.
Failed:   226, Passed:   265

thaystg committed:
Reimplement call function on, and implement set values.
Failed:   192, Passed:   299

thaystg committed:
Failed:   192, Passed:   299

thaystg committed:
Fixing valuetype with null values.

Failed:   184, Passed:   307

thaystg committed:
Implemented Evaluate expressions, conditional breakpoints, all breakpoints tests are passing.
Failed:   172, Passed:   319

thaystg committed:
Fixing evaluate with value type.
Failed:   156, Passed:   335

thaystg committed:
Trim part and add cache.
Failed:   148, Passed:   343

thaystg committed:
Fixing evaluate expression.
Failed:    99, Passed:   392

thaystg committed:
GetPropertiesTests working.

Failed:    53, Passed:   438

thaystg committed:
Passing delegate tests.
Failed:    31, Passed:   460

thaystg committed:
Removing unused code.

thaystg committed:
Implementing exception handler.
Failed:    30, Passed:   461

thaystg committed:
Fixing cfo returning array.
Removing more code.
Removing 2 tests that tests functions which does not exist anymore.
Failed:    18, Passed:   471

thaystg committed:
Fix CallFunctionOn returning primitive types and null.

Failed:     9, Passed:   480

thaystg committed:
Failed:     7, Passed:   482

thaystg committed:
Fixing some tests.
Failed:     2, Passed:   488

thaystg committed:
Removing a lot of code.
Failed:     4, Passed:   485

thaystg committed:
0 ERRORS!

thaystg committed:
Removing more code.
No errors.

thaystg committed:
Merge remote-tracking branch 'upstream/main' into thays_remove_mini_wasm_debugger_and_use_debugger_agent

* upstream/main: (597 commits)
  Fix42292 (#52463)
  [mono] Remove some obsolete emscripten flags. (#53486)
  Fixed path to projects (#53435)
  support ServerCertificateContext in quic (#53175)
  Socket: delete unix local endpoint filename on Close (#52103)
  [mono] Fix sgen_gc_info.memory_load_bytes (#53364)
  Refactor MsQuic's native IP address types. (#53461)
  Re-enabled optimizations for gtFoldExprConst (#53347)
  Add diagnostic support into sample app and AppBuilders on Mono. (#53361)
  Fix issues with virtuals and mdarrays (#53400)
  Split Variant marshalling tests (#53035)
  Update clrjit.natvis to cover GT_SIMD and GT_HWINTRINSIC (#53470)
  remove WSL checks in tests (#53475)
  Always spawn message loop thread for SystemEvents (#53467)
  add AcceptAsync cancellation overloads (#53340)
  Remove unnecessary reference to iOS workload pack in the Mono workload (#53425)
  Add CookieContainer.GetAllCookies (#53441)
  Remove DynamicallyAccessedMembers on JsonSerializer  (#53235)
  [wasm] Re-enable Wasm.Build.Tests (#53433)
  [libraries] Move library tests Feature Switches defaults to Functional tests (#53253)
  ...

thaystg committed:
Fixing added tests.

thaystg committed:
Return javascript callstack after managed callstack.
Step out from managed code return to native wasm or javascript.
Adding debug info to Wasm.Browser.Sample to help testing debugger with sample.

thaystg committed:
Change what Ankit suggested.
Clear cache with valuetypes and pointers after resume or step.

thaystg commented on src/mono/mono/mini/debugger-agent.c:
```diff
@@ -397,6 +367,18 @@ static gint32 suspend_count;
 
+
+#ifndef TARGET_WASM
+#define GET_TLS_DATA(thread) \
```
Yesy

thaystg commented on src/mono/mono/mini/debugger-agent.c:
```diff
@@ -1591,6 +1568,22 @@ static GHashTable *obj_to_objref;
+
+	log_level = log_level;
+	event_requests = g_ptr_array_new ();
+	if (mono_atomic_cas_i32 (&agent_inited, 1, 0) == 1)
```
Yes.

thaystg commented on src/mono/mono/mini/debugger-agent.c:
```diff
@@ -6934,6 +6954,34 @@ vm_commands (int command, int id, guint8 *p, guint8 *end, Buffer *buf)
+		char* assembly_name = decode_string (p, &p, end);
+		//we get 'foo.dll' but mono_assembly_load expects 'foo' so we strip the last dot
+		char *lookup_name = g_strdup (assembly_name);
+		for (i = strlen (lookup_name) - 1; i >= 0; --i) {
+			if (lookup_name [i] == '.') {
+				lookup_name [i] = 0;
+				break;
+			}
+		}
```
I copied this code from mini-wasm-debugger. Probably yes.

thaystg commented on src/mono/mono/mini/debugger-agent.c:
```diff
@@ -6934,6 +6954,34 @@ vm_commands (int command, int id, guint8 *p, guint8 *end, Buffer *buf)
+
+		//resolve the assembly
+		MonoImageOpenStatus status;
+		MonoAssemblyName* aname = mono_assembly_name_new (lookup_name);
```
fixed.

thaystg commented on src/mono/mono/mini/debugger-agent.c:
```diff
@@ -6934,6 +6954,34 @@ vm_commands (int command, int id, guint8 *p, guint8 *end, Buffer *buf)
+			buffer_add_int(buf, -1);
+			break;
+		}
+		mono_assembly_name_free_internal (aname);
```
fixed.

thaystg commented on src/mono/mono/mini/debugger-agent.c:
```diff
@@ -9897,6 +9939,62 @@ wait_for_attach (void)
 }
 
+ErrorCode
+mono_process_dbg_packet (int id, CommandSet command_set, int command, gboolean *no_reply, guint8 *p, guint8 *end, Buffer *buf)
```
done.

thaystg committed:
Fixing suggestions.

thaystg committed:
Fix error on wasm build.

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:
```diff
@@ -0,0 +1,1872 @@
+        mdtBaseType             = 0x72000000,       // Leave this on the high end value. This does not correspond to metadata table
+    }
+
+    internal enum CommandSet {
```
I think we have 3 different enum value casings in this now, is there a reason to avoid using CamelCase for all of them? Is the idea to make the comparison to the C code a little more direct?

thaystg replied:
I just copied from debugger-libs. Changed to CamelCase.

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:
```diff
@@ -0,0 +1,1872 @@
+            return false;
+        }
+
+        public async Task<JObject> CreateJObjectForVariableValue(SessionId sessionId, MonoBinaryReader ret_debugger_cmd_reader, string name, bool isOwn, int typeIdFromAttribute, CancellationToken token)
```
This method is pretty large and doesn't have a lot of shared state in the switch.

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:
```diff
@@ -0,0 +1,1872 @@
+                ret["isOwn"] = true;
+            return ret;
+        }
+        public async Task<bool> IsAsyncMethod(SessionId sessionId, int methodId, CancellationToken token)
```
```suggestion

        public async Task<bool> IsAsyncMethod(SessionId sessionId, int methodId, CancellationToken token)
```

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:
```diff
@@ -0,0 +1,1872 @@
+            return locals;
+
+        }
+        public async Task<JArray> GetValueTypeValues(SessionId sessionId, int valueTypeId, bool accessorPropertiesOnly, CancellationToken token)
```
```suggestion

        public async Task<JArray> GetValueTypeValues(SessionId sessionId, int valueTypeId, bool accessorPropertiesOnly, CancellationToken token)
```

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -460,32 +464,75 @@ protected override async Task<bool> AcceptCommand(MessageId id, string method, J
+            }
+            if (objectId.Scheme == "object" || objectId.Scheme == "valuetype" || objectId.Scheme == "pointer" || objectId.Scheme == "array")
+            {
+                if (objectId.Scheme == "object")
```
use a switch expression?

thaystg replied:
Done.

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -498,50 +545,83 @@ private async Task<bool> OnSetVariableValue(MessageId id, int scopeId, string va
-                // Runtime.callFunctionOn result object
-                string value_json_str = res.Value["result"]?["value"]?["__value_as_json_string__"]?.Value<string>();
-                if (value_json_str != null)
+            //Console.WriteLine($"RuntimeGetProperties - {args}");
```
Something like
```csharp
            //Console.WriteLine($"RuntimeGetProperties - {args}");
            try {
                switch (objectId.Scheme)
                {
                    case "scope":
                        return await GetScopeProperties(id, int.Parse(objectId.Value), token);
                    case "valuetype":
                        return Result.OkFromObject(new { result = await sdbHelper.GetValueTypeValues(id, int.Parse(objectId.Value), accessorPropertiesOnly, token)});
                    case "array":
                        return Result.OkFromObject(new { result = await sdbHelper.GetArrayValues(id, int.Parse(objectId.Value), token)});
                    case "object":
                        return Result.OkFromObject(new { result = await sdbHelper.GetObjectValues(id, int.Parse(objectId.Value), true, false, accessorPropertiesOnly, ownProperties, token) });
                    case "pointer":
                        return Result.OkFromObject(new { result = await sdbHelper.GetPointerContent(id, int.Parse(objectId.Value), token)});
                    case "cfo_res":
                    {
                        Result res = await SendMonoCommand(id, MonoCommands.GetDetails(int.Parse(objectId.Value), args), token);
                        // Runtime.callFunctionOn result object
                        string value_json_str = res.Value["result"]?["value"]?["__value_as_json_string__"]?.Value<string>();
                        return Result.OkFromObject(new { result = value_json_str != null ? (object) JArray.Parse(value_json_str) : new {} });
                    }
                    default:
                        return Result.Err($"Unable to RuntimeGetProperties '{objectId}'");
                }
            }
            catch (Exception e) {
                return Result.Err($"Unable to RuntimeGetProperties '{objectId}' - {e}");
            }
```

here instead of the repeated ifs?

thaystg replied:
Done.

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -498,50 +545,83 @@ private async Task<bool> OnSetVariableValue(MessageId id, int scopeId, string va
         }
 
-        private async Task<Result> RuntimeGetProperties(MessageId id, DotnetObjectId objectId, JToken args, CancellationToken token)
+        internal async Task<Result> RuntimeGetProperties(SessionId id, DotnetObjectId objectId, JToken args, CancellationToken token)
```
It feels a little odd to return a result from something that doesn't start with a message.  It looks like this was needed  to make it callable from MemberReferenceResolver that then unpacks the result?

thaystg replied:
Changed.

lewing reviewed (commented):
This is looking great but I have some style questions while I keep reviewing.  There are a few small formatting things to fix as well but I stopped pointing them out individually.

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:
```diff
@@ -0,0 +1,1872 @@
+            var ret_debugger_cmd_reader = await SendDebuggerAgentCommand(sessionId, (int) CommandSet.EVENT_REQUEST, (int) CmdEventRequest.SET, command_params, token);
+            return true;
+        }
+        internal async Task<MonoBinaryReader> SendDebuggerAgentCommand(SessionId sessionId, int command_set, int command, MemoryStream parms, CancellationToken token)
```
I was thinking something like         

```csharp
internal Task<MonoBinaryReader> SendDebuggerAgentCommand<T>(SessionId sessionId, T command, MemoryStream parms, CancellationToken token) where T : Enum
        {
            CommandSet set = command switch {
                CmdAppDomain => CommandSet.APPDOMAIN,
                CmdArray => CommandSet.ARRAY_REF,
                CmdVM => CommandSet.VM,
                CmdEventRequest => CommandSet.EVENT_REQUEST,
                CmdMethod => CommandSet.METHOD,
                CmdModule => CommandSet.MODULE,
                CmdAssembly => CommandSet.ASSEMBLY,
                _ => throw new Exception ("Unknown CommandSet")
            };
            return SendDebuggerAgentCommand(sessionId, (int)set, (int)(object)command, parms, token);
        }
```

Might clean things up a bit

thaystg replied:
Done.

thaystg commented on src/mono/mono/mini/debugger-agent.c:
```diff
@@ -3586,13 +3603,15 @@ process_event (EventKind event, gpointer arg, gint32 il_offset, MonoContext *ctx
 		}
 		case EVENT_KIND_USER_BREAK: {
+#ifndef TARGET_WASM
 			DebuggerTlsData *tls;
 			tls = (DebuggerTlsData *)mono_native_tls_get_value (debugger_tls_id);
 			g_assert (tls);
 			// We are already processing a breakpoint event
 			if (tls->disable_breakpoints)
 				return;
 			mono_stopwatch_stop (&tls->step_time);
```
I will remove mono_stop_timer_watch and added a macro to GET_DEBUGGER_TLS, and I used it in both places.

thaystg committed:
Apply suggestions from code review

Co-authored-by: Larry Ewing <lewing@microsoft.com>

thaystg committed:
Changing what was suggested by @lewing.

thaystg committed:
Fix pointer tests.

thaystg committed:
Refactoring CreateJObjectForVariableValue

thaystg committed:
Changing what @lewing suggested.

lewing reviewed (commented):
More style thoughts, but feel free to ignore

thaystg committed:
Apply suggestions from code review

Co-authored-by: Larry Ewing <lewing@microsoft.com>

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs

Co-authored-by: Larry Ewing <lewing@microsoft.com>

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs

Co-authored-by: Larry Ewing <lewing@microsoft.com>

thaystg committed:
Fixing @lewing changes.

filipnavara commented:
The amount of this error in the Helix logs is worrying:
`/__w/1/s/src/mono/mono/metadata/mono-hash.c:282: assertion 'hash != NULL' failed`
(eg. https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-52300-merge-bf5b213642ce4e2dae/System.Runtime.Extensions.Tests/console.cd1b4bde.log?sv=2019-07-07&se=2021-07-07T03%3A41%3A46Z&sr=c&sp=rl&sig=nnN1S1pF0TwQdFeeCEwGR15Fivujlty4Gl4wTHGmgiY%3D)

thaystg commented:
Thanks @filipnavara I'll take a look! :)

thaystg committed:
Trying to fix CI.

lewing reviewed (approved):
Let's do a lot of manual testing now.

cc @radical @pavelsavara @radekdoulik @kg @Daniel-Genkin-MS

Daniel-Genkin commented:
I just did some testing on Windows and found the following behaviours. Although I'm not sure if they are directly applicable to this PR. Although other than that, this works really well and is really cool!!

- if you hover your mouse over the html tags in the elements tab of chrome's devtools then the debugger's log prints the following error:
`fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] sending error response for id: msg-:::180 -> [Result: IsOk: False, IsErr: True, Value: , Error: {   "code": -32000,   "message": "Could not compute box model." } ]`

- if you "Deactivate all breakpoints" and refresh the page being debugged the breakpoints set in the native code are lost. So when you enable them, the breakpoints aren't re-added. However if you manually disable all the breakpoints, they aren't lost upon page refreshes.

- when cancelling the debugger via CTRL + C, it prints an error stack trace to the debugger's log
`fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] DevToolsProxy::Run: Exception System.AggregateException: One or more errors occurred. (A task was canceled.)  ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.    at System.Threading.Tasks.Task.GetExceptions(Boolean includeTaskCanceledExceptions)    at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)    at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)    at Microsoft.WebAssembly.Diagnostics.DevToolsProxy.Run(Uri browserUri, WebSocket ideSocket) in C:\runtime\src\mono\wasm\debugger\BrowserDebugProxy\DevToolsProxy.cs:line 283    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task`1.TrySetResult(TResult result)    at System.Threading.Tasks.TaskFactory.CompleteOnInvokePromise.Invoke(Task completingTask)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task.CancellationCleanupLogic()    at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)    at Microsoft.WebAssembly.Diagnostics.DevToolsProxy.ReadOne(WebSocket socket, CancellationToken token) in C:\runtime\src\mono\wasm\debugger\BrowserDebugProxy\DevToolsProxy.cs:line 61    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task.CancellationCleanupLogic()    at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)    at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1.SetException(Exception exception)    at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task.CancellationCleanupLogic()    at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)    at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder.SetException(Exception exception)    at System.Net.WebSockets.ManagedWebSocket.EnsureBufferContainsAsync(Int32 minimumRequiredBytes, CancellationToken cancellationToken, Boolean throwOnPrematureClosure)    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task.CancellationCleanupLogic()    at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)    at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1.SetException(Exception exception)    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 destination, CancellationToken cancellationToken)    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task.CancellationCleanupLogic()    at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)    at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1.SetException(Exception exception)    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1UpgradeMessageBody.ReadAsyncInternalAwaited(ValueTask`1 readTask, CancellationToken cancellationToken)    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.ThreadPoolWorkQueue.Dispatch()    at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() --- End of stack trace from previous location ---     --- End of inner exception stack trace ---    at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)    at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)    at Microsoft.WebAssembly.Diagnostics.DevToolsProxy.Run(Uri browserUri, WebSocket ideSocket) in C:\runtime\src\mono\wasm\debugger\BrowserDebugProxy\DevToolsProxy.cs:line 283
Terminate batch job (Y/N)?`

lewing commented:
@Daniel-Genkin thank you for the testing

> I just did some testing on Windows and found the following behaviours. Although I'm not sure if they are directly applicable to this PR. Although other than that, this works really well and is really cool!!
> 
> * if you hover your mouse over the html tags in the elements tab of chrome's devtools then the debugger's log prints the following error:
>   `fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] sending error response for id: msg-:::180 -> [Result: IsOk: False, IsErr: True, Value: , Error: {   "code": -32000,   "message": "Could not compute box model." } ]`

This needs to have a test added and needs to be fixed. I don't consider it blocking for landing this pr but it is blocking for net6.0

> * if you "Deactivate all breakpoints" and refresh the page being debugged the breakpoints set in the native code are lost. So when you enable them, the breakpoints aren't re-added. However if you manually disable all the breakpoints, they aren't lost upon page refreshes.

This sounds like https://github.com/dotnet/aspnetcore/issues/27234

thaystg commented:
> * if you hover your mouse over the html tags in the elements tab of chrome's devtools then the debugger's log prints the following error:
>   `fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] sending error response for id: msg-:::180 -> [Result: IsOk: False, IsErr: True, Value: , Error: {   "code": -32000,   "message": "Could not compute box model." } ]`

This also happens in the version without my change, we need to investigate if debugger proxy is returning this error or the browser is returning this error.

thaystg commented:
> * when cancelling the debugger via CTRL + C, it prints an error stack trace to the debugger's log
>   `fail: Microsoft.WebAssembly.Diagnostics.DevToolsProxy[0] DevToolsProxy::Run: Exception System.AggregateException: One or more errors occurred. (A task was canceled.)  ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.    at System.Threading.Tasks.Task.GetExceptions(Boolean includeTaskCanceledExceptions)    at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)    at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)    at Microsoft.WebAssembly.Diagnostics.DevToolsProxy.Run(Uri browserUri, WebSocket ideSocket) in C:\runtime\src\mono\wasm\debugger\BrowserDebugProxy\DevToolsProxy.cs:line 283    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task`1.TrySetResult(TResult result)    at System.Threading.Tasks.TaskFactory.CompleteOnInvokePromise.Invoke(Task completingTask)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task.CancellationCleanupLogic()    at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)    at Microsoft.WebAssembly.Diagnostics.DevToolsProxy.ReadOne(WebSocket socket, CancellationToken token) in C:\runtime\src\mono\wasm\debugger\BrowserDebugProxy\DevToolsProxy.cs:line 61    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task.CancellationCleanupLogic()    at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)    at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1.SetException(Exception exception)    at System.Net.WebSockets.ManagedWebSocket.ReceiveAsyncPrivate[TWebSocketReceiveResultGetter,TWebSocketReceiveResult](Memory`1 payloadBuffer, CancellationToken cancellationToken, TWebSocketReceiveResultGetter resultGetter)    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task.CancellationCleanupLogic()    at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)    at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder.SetException(Exception exception)    at System.Net.WebSockets.ManagedWebSocket.EnsureBufferContainsAsync(Int32 minimumRequiredBytes, CancellationToken cancellationToken, Boolean throwOnPrematureClosure)    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task.CancellationCleanupLogic()    at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)    at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1.SetException(Exception exception)    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 destination, CancellationToken cancellationToken)    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)    at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)    at System.Threading.Tasks.Task.CancellationCleanupLogic()    at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)    at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1.SetException(Exception exception)    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1UpgradeMessageBody.ReadAsyncInternalAwaited(ValueTask`1 readTask, CancellationToken cancellationToken)    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)    at System.Threading.ThreadPoolWorkQueue.Dispatch()    at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() --- End of stack trace from previous location ---     --- End of inner exception stack trace ---    at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)    at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)    at Microsoft.WebAssembly.Diagnostics.DevToolsProxy.Run(Uri browserUri, WebSocket ideSocket) in C:\runtime\src\mono\wasm\debugger\BrowserDebugProxy\DevToolsProxy.cs:line 283 Terminate batch job (Y/N)?`

This one I think that always happens without my change, but probably we should also try to fix.

thaystg commented:
@Daniel-Genkin thanks for testing!!! :)

thaystg commented:
Please, if anyone find anything, please comment here and I will fix and change what is suggested in another PR.
Thanks a lot!
